File: src/conviso_mcp/feed_gateway.py
from graphql_client import GraphQLClient
import dotenv, os

dotenv.load_dotenv()

class FeedGateway:
    def __init__(self, base_url, graphql_api_key: str = ""):
        self.base_url = base_url

        graphql_api_key = os.environ.get("CONVISO_API_KEY")

        self.graphql = GraphQLClient(f"{self.base_url}/graphql", graphql_api_key)
        pass

    @staticmethod
    def get_base_url() -> str:
        dotenv.load_dotenv()

        ENV = os.getenv("STAGING", False)
        if not ENV:
            base_url = "https://app.convisoappsec.com"
        else:
            base_url = "https://staging.convisoappsec.com"

        return base_url


    def get_companies(self, page: int = 1, limit: int = 10, search=""):
        companies = self.graphql.get_companies(page=page, limit=limit, search=search)
        return companies
    
    def get_company_by_id(self, company_id: int):
        return self.graphql.get_company_by_id(company_id)

    def get_issue_by_id(self, issue_id: str, return_snippets: bool = False):
        """Get details of a issue in Conviso Platform"""
        issue = self.graphql.get_issue_by_id(issue_id, return_snippets=return_snippets)
        return issue

    def get_issues(self, company_id: str, search: str, page: int = 1, limit: int = 1, project_id: int = None):
        issues = self.graphql.get_issues(company_id, search=search, page=page, limit=limit, project_id=project_id)
        return issues

    def get_issue_with_company_id(self, company_id: str, issue_id: int):
        issues = self.graphql.get_issues(company_id, search="", page=1, limit=5, issue_ids=[issue_id])
        return issues
    
    def get_issues_by_asset_ids(self, company_id: str, page: int = 1, limit: int = 1, asset_ids = []):
        issues = self.graphql.get_issues(company_id, search="", page=page, limit=limit, asset_ids=asset_ids)
        return issues

    def get_projects(self, company_id: int, page: int = 1, limit: int = 1000, search: str = ""):
        projects = self.graphql.get_projects(company_id, page=page, limit=limit, search=search)
        return projects
    
    def get_project_by_id(self, project_id: int):
        projects = self.graphql.get_project_by_id(project_id)
        return projects
    
    def get_assets(self, company_id: int, page: int = 1, limit: int = 1000):
        return self.graphql.get_assets_by_company(company_id=company_id, page=page, limit=limit)
    
    def get_asset_by_id(self, asset_id: int):
        return self.graphql.get_asset_by_id(asset_id=asset_id)
    
    def get_top_vulnerabilities(self, company_id: int):
        return self.graphql.get_top_vulnerabilities(company_id)
    
    def generate_project_report(
        self,
        project_id: int,
        language: str = "en",
        vulnerability_criticity: list = None,
        vulnerability_statuses: list = None,
        requirements: bool = True,
        evidences: bool = True
    ):
        return self.graphql.generate_project_report(project_id, language, vulnerability_criticity, vulnerability_statuses, requirements, evidences)

    def generate_project_report_progress(self, project_id: int, report_id: int):
        return self.graphql.generate_project_report_progress(project_id, report_id)

    def create_project_url(self, company_id: int, project_id: int):
        return f"{self.base_url}/spa/company/{company_id}/projects/{project_id}"

    def create_issue_url(self, company_id: int, issue_id: int):
        return f"{self.base_url}/spa/company/{company_id}/vulnerabilities?title=&search={issue_id}"
    

if __name__ == "__main__":
    print(FeedGateway("https://staging.convisoappsec.com").get_issue_with_company_id(248,135606)["issues"]["collection"][0]["id"])

File: node/src/conviso_mcp/feed_gateway.js
import 'dotenv/config';
import { GraphQLClient } from './graphql_client.js';
import { listMutations, describeMutation } from './mutations.js';

class FeedGateway {
  constructor(graphql_api_key = '') {
    this.base_url = 'https://app.convisoappsec.com';
    const apiKey = process.env.CONVISO_API_KEY || graphql_api_key || '';
    this.graphql = new GraphQLClient(`${this.base_url}/graphql`, apiKey);
  }

  // search = name contains; label_eq = exact name match.
  async get_companies(page = 1, limit = 10, search = '', label_eq = null) {
    return this.graphql.get_companies(page, limit, search, label_eq);
  }

  async get_company_by_id(company_id) {
    return this.graphql.get_company_by_id(company_id);
  }

  async get_issue_by_id(issue_id, return_snippets = false) {
    return this.graphql.get_issue_by_id(issue_id, return_snippets);
  }

  // Single rich issues passthrough (parity with Python gateway.get_issues kwargs).
  async getIssues(company_id, opts = {}) {
    return this.graphql.getIssues(company_id, opts);
  }

  async get_issue_with_company_id(company_id, issue_id) {
    return this.graphql.get_issues(company_id, '', 1, 5, null, [issue_id]);
  }

  async get_issues_by_asset_ids(company_id, page = 1, limit = 10, asset_ids = [], search = '', opts = {}) {
    return this.graphql.getIssues(company_id, {
      ...opts,
      page,
      limit,
      search,
      assetIds: asset_ids,
    });
  }

  async get_projects(company_id, page = 1, limit = 1000, search = '', opts = {}) {
    return this.graphql.get_projects(company_id, page, limit, search, opts);
  }

  async get_project_by_id(project_id) {
    return this.graphql.get_project_by_id(project_id);
  }

  async get_assets(company_id, page = 1, limit = 1000, opts = {}) {
    return this.graphql.get_assets_by_company(company_id, page, limit, opts);
  }

  async get_asset_by_id(asset_id) {
    return this.graphql.get_asset_by_id(asset_id);
  }

  async get_top_vulnerabilities(company_id, opts = {}) {
    return this.graphql.get_top_vulnerabilities(company_id, opts);
  }

  async generate_project_report(project_id, language = 'en', vulnerability_criticity = null, vulnerability_statuses = null, requirements = true, evidences = true) {
    return this.graphql.generate_project_report(project_id, language, vulnerability_criticity, vulnerability_statuses, requirements, evidences);
  }

  async generate_project_report_progress(project_id, report_id) {
    return this.graphql.generate_project_report_progress(project_id, report_id);
  }

  create_project_url(company_id, project_id) {
    return `${this.base_url}/spa/company/${company_id}/projects/${project_id}`;
  }

  create_issue_url(company_id, issue_id) {
    return `${this.base_url}/spa/company/${company_id}/vulnerabilities?title=&search=${issue_id}`;
  }

  async get_mttr_over_time(company_id, start_date, end_date, severities = null, statuses = null, asset_ids = null, asset_tags = null) {
    return this.graphql.get_mttr_over_time(company_id, start_date, end_date, severities, statuses, asset_ids, asset_tags);
  }

  async get_overall_risk_score_history(company_id) {
    return this.graphql.get_overall_risk_score_history(company_id);
  }

  // --- Mutations -------------------------------------------------------------

  // Discovery/introspection are pure (no network) — answered straight from the catalog.
  list_mutations(opts = {}) {
    return listMutations(opts);
  }

  describe_mutation(name) {
    return describeMutation(name);
  }

  async execute_mutation(name, variables = {}, returnFields = null) {
    return this.graphql.executeMutation(name, variables, returnFields);
  }

  async change_issue_status(a = {}) {
    return this.graphql.change_issue_status(a);
  }

  async create_source_code_vulnerability(a = {}) {
    return this.graphql.create_source_code_vulnerability(a);
  }

  async create_project(a = {}) {
    return this.graphql.create_project(a);
  }

  async create_asset(a = {}) {
    return this.graphql.create_asset(a);
  }

  async create_ticket(a = {}) {
    return this.graphql.create_ticket(a);
  }

  async run_dast(a = {}) {
    return this.graphql.run_dast(a);
  }

  async trigger_pentest(a = {}) {
    return this.graphql.trigger_pentest(a);
  }

  async create_pentest_artifact(a = {}) {
    return this.graphql.create_pentest_artifact(a);
  }

  // --- Reads (curated query tools) -------------------------------------------

  async get_tickets(company_id, opts = {}) {
    return this.graphql.get_tickets(company_id, opts);
  }

  async get_ticket(company_id, ticket_id) {
    return this.graphql.get_ticket(company_id, ticket_id);
  }

  async get_requirements(scope_id, opts = {}) {
    return this.graphql.get_requirements(scope_id, opts);
  }

  async get_requirement(company_id, requirement_id) {
    return this.graphql.get_requirement(company_id, requirement_id);
  }

  async get_project_requirements(project_id) {
    return this.graphql.get_project_requirements(project_id);
  }

  async get_applications(company_id, search = null) {
    return this.graphql.get_applications(company_id, search);
  }

  async get_application(company_id, application_id) {
    return this.graphql.get_application(company_id, application_id);
  }

  async get_scan_histories(company_id, opts = {}) {
    return this.graphql.get_scan_histories(company_id, opts);
  }

  async get_asset_scans_count(company_id) {
    return this.graphql.get_asset_scans_count(company_id);
  }

  async get_sbom_components(company_id, opts = {}) {
    return this.graphql.get_sbom_components(company_id, opts);
  }

  async get_pentest_artifacts(company_id, opts = {}) {
    return this.graphql.get_pentest_artifacts(company_id, opts);
  }

  async get_pentest_artifact(artifact_id) {
    return this.graphql.get_pentest_artifact(artifact_id);
  }

  async get_pentest_execution(execution_id) {
    return this.graphql.get_pentest_execution(execution_id);
  }

  async get_threat_model_artifacts(company_id, opts = {}) {
    return this.graphql.get_threat_model_artifacts(company_id, opts);
  }

  async get_threat_model_artifact(artifact_id) {
    return this.graphql.get_threat_model_artifact(artifact_id);
  }
}

export { FeedGateway };
